lib/AST/ProtocolConformance.cpp

```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
+        return false;
+    return true;
+  }
+  case ProtocolConformanceKind::Inherited: {
+    // Substitute the base.
+    auto inheritedConformance
+      = cast<InheritedProtocolConformance>(this);
+    return inheritedConformance->getInheritedConformance()->isCanonical();
+  }
+  case ProtocolConformanceKind::Specialized: {
+    // Substitute the substitutions in the specialized conformance.
+    auto spec = cast<SpecializedProtocolConformance>(this);
+    auto genericConformance = spec->getGenericConformance();
+    if (!genericConformance->isCanonical())
+      return false;
+    auto specSubs = spec->getGenericSubstitutions();
+    for (auto sub : specSubs) {
+      if (!sub.getReplacement()->isCanonical())
+        return false;
+      for (auto conf : sub.getConformances()) {
+        if (!conf.isCanonical())
+          return false;
+      }
+    }
+    return true;
+  }
+  }
+  llvm_unreachable("bad ProtocolConformanceKind");
+}
+
+Substitution Substitution::getCanonicalSubstitution(bool *wasCanonical) const {
+  bool createdNewCanonicalConformances = false;
+  bool createdCanReplacement = false;
+  SmallVector<ProtocolConformanceRef, 4> newCanConformances;
+
+  CanType canReplacement = getReplacement()->getCanonicalType();
+
+  if (!getReplacement()->isCanonical()) {
+    createdCanReplacement = true;
+  }
+
+  for (auto conf : getConformances()) {
+    if (conf.isCanonical()) {
+      newCanConformances.push_back(conf);
+      continue;
+    }
+    newCanConformances.push_back(conf.getCanonicalConformanceRef());
+    createdNewCanonicalConformances = true;
+  }
+
+  ArrayRef<ProtocolConformanceRef> canConformances = getConformances();
+  if (createdNewCanonicalConformances) {
+    auto &C = canReplacement->getASTContext();
+    canConformances = C.AllocateCopy(newCanConformances);
+  }
+
+  if (createdCanReplacement || createdNewCanonicalConformances) {
+    if (wasCanonical)
+      *wasCanonical = false;
+    return Substitution(canReplacement, canConformances);
+  }
+  if (wasCanonical)
+    *wasCanonical = true;
+  return *this;
+}
+
+SubstitutionList swift::getCanonicalSubstitutionList(SubstitutionList subs) {
+  SmallVector<Substitution, 4> newCanSubstitutions;
+  bool subListWasCanonical = true;
+  for (auto &sub : subs) {
+    bool subWasCanonical = false;
+    auto canSub = sub.getCanonicalSubstitution(&subWasCanonical);
+    if (!subWasCanonical)
+      subListWasCanonical = false;
+    newCanSubstitutions.push_back(canSub);
+  }
+  if (subListWasCanonical)
+    return subs;
+  return newCanSubstitutions[0].getReplacement()->getASTContext().AllocateCopy(
+      newCanSubstitutions);
+}
+
+/// Check of all types used by the conformance are canonical.
+ProtocolConformance *ProtocolConformance::getCanonicalConformance() {
+  if (isCanonical())
+    return this;
+
+  switch (getKind()) {
+  case ProtocolConformanceKind::Normal: {
+    //assert(isCanonical() && "Normal conformances should always be canonical");
```

Comment:
The NormalConformances are canonical by construction (see how NormalProtocolConformance::Profile is implemented. It uses only the protocol for hashing, everything else is derived from it), but the NormalConformance::getType and NormalConformance::getInterfaceType are not necessarily canonical, because they are derived and do no participate in profiling. Thus we cannot really  leave this check in.